sonic-stormond/scripts/stormond

```diff
@@ -0,0 +1,374 @@
+
+        # If the daemon is re-init-ing after a planned reboot or powercycle, there would be no storage info 
+        # in the STATE_DB. Therefore, we would need to parse the total and hitherto latest procfs reads 
+        # and writes from the FSIO JSON file and use those reads/writes values as a baseline.
+        elif self.use_statedb_baseline == False and self.use_fsio_json_baseline == True:
+            fsio_dict["total_fsio_reads"] = str(int(self.fsio_rw_json[device]["total_fsio_reads"]) + int(fsio_dict["latest_fsio_reads"]))
+            fsio_dict["total_fsio_writes"] = str(int(self.fsio_rw_json[device]["total_fsio_writes"]) + int(fsio_dict["latest_fsio_writes"]))
+
+        # The only scenario where there would be storage info present in the STATE_DB is when the daemon is 
+        # coming back up after a crash.
+
+        # In this scenario, we use the STATE_DB values as the SoT. We use the 'latest_fsio_reads/writes'
+        # values from STATE_DB, which is the values from the last invocation of get_fs_io_reads/writes 
+        # on the storage disk that was posted to STATE_DB, and the values obtained from the most recent
+        # invocation of get_fs_io_reads/writes (prior to this function being called) to determine the 
+        # additional procfs reads and writes that have happened on the FS while the daemon was down. 
+        
+        # We then add these additional values to the previous values of total_fsio_reads/writes to 
+        # determine the new total procfs reads/writes.
+
+        elif self.use_statedb_baseline == True:
+            additional_procfs_reads = int(fsio_dict["latest_fsio_reads"]) - int(self.fsio_rw_statedb[device]["latest_fsio_reads"])
+            additional_procfs_writes = int(fsio_dict["latest_fsio_writes"]) - int(self.fsio_rw_statedb[device]["latest_fsio_writes"])
+
+            fsio_dict["total_fsio_reads"] = str(int(self.fsio_rw_statedb[device]["total_fsio_reads"]) + additional_procfs_reads)
+            fsio_dict["total_fsio_writes"] = str(int(self.fsio_rw_statedb[device]["total_fsio_writes"]) + additional_procfs_writes)
+        
+        return fsio_dict["total_fsio_reads"], fsio_dict["total_fsio_writes"]
+
+
+
+    # Update the Storage device info to State DB
+    def update_storage_info_status_db(self, disk_device, disk_fields, kvp_dict):
+
+        fvp = swsscommon.FieldValuePairs([(field, str(value)) for field, value in kvp_dict.items()])
+        self.device_table.set(disk_device, fvp)
+
+
+    # Get Static attributes and update the State DB, once
+    def get_static_fields(self):
+
+        # Get relevant information about each Storage Device on the switch
+        for storage_device, storage_object in self.storage.devices.items():
+            try:
+                # Unlikely scenario
+                if storage_object is None:
+                    self.log_error("{} does not have an instantiated object. Static Information cannot be gathered.".format(storage_device))
```

Comment:
On simulated devices this error happens if there is a different type of storage. For eg if a server has cd rom this error is seen.

```
brw-rw---- 1 root cdrom 11, 0 May 25 18:28 /dev/sr0
root@r-panther-simx-03:~# ls -l /dev/sda
brw-rw---- 1 root disk 8, 0 May 25 18:28 /dev/sda
root@r-panther-simx-03:~# ls -l /sys/block/
total 0
lrwxrwxrwx 1 root root 0 May 25 18:29 loop0 -> ../devices/virtual/block/loop0
lrwxrwxrwx 1 root root 0 May 25 18:29 loop1 -> ../devices/virtual/block/loop1
lrwxrwxrwx 1 root root 0 May 25 18:29 loop2 -> ../devices/virtual/block/loop2
lrwxrwxrwx 1 root root 0 May 25 18:29 loop3 -> ../devices/virtual/block/loop3
lrwxrwxrwx 1 root root 0 May 25 18:29 loop4 -> ../devices/virtual/block/loop4
lrwxrwxrwx 1 root root 0 May 25 18:29 loop5 -> ../devices/virtual/block/loop5
lrwxrwxrwx 1 root root 0 May 25 18:29 loop6 -> ../devices/virtual/block/loop6
lrwxrwxrwx 1 root root 0 May 25 18:29 loop7 -> ../devices/virtual/block/loop7
lrwxrwxrwx 1 root root 0 May 25 18:28 sda -> ../devices/pci0000:00/0000:00:06.0/ata3/host2/target2:0:0/2:0:0:0/block/sda
lrwxrwxrwx 1 root root 0 May 25 18:29 sr0 -> ../devices/pci0000:00/0000:00:01.1/ata1/host0/target0:0:0/0:0:0:0/block/sr0
```

The following error will be seen. Request if this can be gracefully handled
```
2024 May 24 08:28:30.264696 r-panther-simx-03 ERR pmon#stormond[26]: sr0 does not have an instantiated object. Static Information cannot be gathered.
2024 May 24 08:28:30.265152 r-panther-simx-03 ERR pmon#stormond[26]: Storage device 'sr0' does not have an instantiated object. Dynamic Information cannot be gathered.

```

Reply:
This is seen because the StorageDevices class has not instantiated a class object for this device type. This is because we determine what the object type should be in the [object factory function](https://github.com/sonic-net/sonic-platform-common/blob/0b95457ee1da9e61d62091fddf4ceaced90f4266/sonic_platform_base/sonic_storage/storage_devices.py#L55) by searching for certain keywords in the realpath of the device. Since 'sr0' does not match any condition, no object is created.

That said, the code should have created SsdUtil object (and you should have seen static/dynamic logs) for the `sda` device at `../devices/pci0000:00/0000:00:06.0/ata3/host2/target2:0:0/2:0:0:0/block/sda` -- was this not the case?

Reply:
Resolved per our discussion offline.